### PR TITLE
Local TVM calculates gas

### PR DESCRIPTION
**Pull request by AlexeyVavilin**

### CODE REVIEW

File: ton_client/client/src/contracts/run.rs [hunk 1]
```diff
@@ -156,7 +173,7 @@ pub(crate) fn run(_context: &mut ClientContext, params: ParamsOfRun) -> ApiResul
     }
 }
 
-pub(crate) fn local_run(_context: &mut ClientContext, params: ParamsOfLocalRun) -> ApiResult<ResultOfRun> {
+pub(crate) fn local_run(_context: &mut ClientContext, params: ParamsOfLocalRun) -> ApiResult<ResultOfLocalRun> {
```
Comment (melsomino) [inline ton_client/client/src/contracts/run.rs]: If you have started use `_context` parameter you should rename it to `context`

File: ton_sdk/src/contract.rs [hunk 2]
```diff
@@ -14,15 +14,16 @@
 
 use crate::*;
 use ed25519_dalek::{Keypair, PublicKey};
-use std::convert::Into;
+use chrono::prelude::*;
```
Comment (melsomino) [inline ton_sdk/src/contract.rs]: It is bad practice to use `*` import

File: ton_sdk/src/contract.rs [hunk 3]
```diff
@@ -432,24 +453,45 @@ impl Contract {
         Ok(acc)
     }
 
+
+
     /// Invokes local TVM instance with provided inbound message.
     /// Returns outbound messages generated by contract function
-    pub fn local_call(&self, message: TvmMessage) -> SdkResult<Vec<Message>> {
+    pub fn local_call(&self, message: TvmMessage) -> SdkResult<(Vec<Message>, u128)> {
```
Comment (melsomino) [inline ton_sdk/src/contract.rs]: 1) Suggest replace resulting tuple with struct:
struct LocalCallResult { messages, gas_fee }
2) Doc comment doesn't reflect new result structure

File: ton_sdk/src/contract.rs [hunk 4]
```diff
@@ -696,4 +738,18 @@ impl Contract {
 
         Ok(TvmMessage::construct_from(&mut root_cells.remove(0).into())?)
     }
+
+    fn balance_other_as_hasmape(&self) -> SdkResult<HashmapE> {
```
Comment (melsomino) [inline ton_sdk/src/contract.rs]: hasmape? `h` missing?

File: ton_sdk/src/contract.rs [hunk 5]
```diff
@@ -432,24 +453,45 @@ impl Contract {
...
+        } else {
+            gas_used * GAS_PRICE_WC
+        };
+
+        Ok((messages, gas_fee))
     }
 
     /// Invokes local TVM instance with provided inbound message.
     /// Returns outbound messages generated by contract function
     pub fn local_call_json(&self, func: String, input: String, abi: String, key_pair: Option<&Keypair>)
```
Comment (melsomino) [inline ton_sdk/src/contract.rs]: suggest to use a struct instead of tuple

File: ton_sdk/src/local_tvm.rs [hunk 6]
```diff
@@ -14,29 +14,38 @@
...
+use ton_vm::executor::gas::gas_state::Gas;
 
 #[cfg(test)]
 #[path = "tests/test_local_tvm.rs"]
 mod tests;
 
 #[allow(dead_code)]
-pub fn local_contract_call(code: Arc<CellData>, data: Option<Arc<CellData>>, msg: &Message)
--> SdkResult<Vec<Message>> {
+pub fn local_contract_call(
```
Comment (melsomino) [inline ton_sdk/src/local_tvm.rs]: Too many parameters. I suggest to introduce struct to represent parameters.

File: ton_sdk/src/local_tvm.rs [hunk 7]
```diff
@@ -14,29 +14,38 @@
...
+    balance: u128,
+    balance_other: HashmapE,
+    address: &MsgAddressInt,
+    config_params: Option<Arc<CellData>>,
+    timestamp: u32,
+    code: Arc<CellData>,
+    data: Option<Arc<CellData>>,
+    msg: &Message)
+-> SdkResult<(Vec<Message>, i64)> {
```
Comment (melsomino) [inline ton_sdk/src/local_tvm.rs]: tuple again

File: ton_sdk/src/tests/test_local_tvm.rs [hunk 8]
```diff
@@ -34,6 +34,7 @@ fn test_local_piggy_call() {
         false)
             .expect("Error decoding result");
 
+    println!("Gas fee {}", gas_fee);
```
Comment (melsomino) [inline ton_sdk/src/tests/test_local_tvm.rs]: I haven't seen any assertions related to `gas_fee` (print only).

Comment (AlexeyVavilin) [reply ton_sdk/src/tests/test_local_tvm.rs]: Such test is implemented in ton-client-js

File: ton_sdk/src/contract.rs (hunk 2 above)
Comment (AlexeyVavilin) [reply ton_sdk/src/contract.rs]: It is code from example in crate documentation.
https://crates.io/crates/chrono

File: ton_sdk/src/local_tvm.rs (hunk 6 above)
Comment (AlexeyVavilin) [reply ton_sdk/src/local_tvm.rs]: It is not a good idea to pack parameters into struct just to decrease function header. Struct should represent some logical object but these parameters are independent values

File: ton_sdk/src/local_tvm.rs (hunk 7 above)
Comment (AlexeyVavilin) [reply ton_sdk/src/local_tvm.rs]: Ley it be tuple here. Made struct in wrapper functions

File: ton_client/client/src/contracts/run.rs (hunk 1 above)
Comment (AlexeyVavilin) [reply ton_client/client/src/contracts/run.rs]: Done

File: ton_sdk/src/contract.rs (hunk 3 above)
Comment (AlexeyVavilin) [reply ton_sdk/src/contract.rs]: Done

File: ton_sdk/src/contract.rs (hunk 4 above)
Comment (AlexeyVavilin) [reply ton_sdk/src/contract.rs]: Done

**Pull request merged**